src/coreclr/jit/optimizemaskconversions.cpp

```diff
@@ -0,0 +1,508 @@
+        bool                 isLocalStore     = false;
+        bool                 isLocalUse       = false;
+        bool                 addConversion    = false;
+        bool                 removeConversion = false;
+
+        if ((*use)->OperIs(GT_STORE_LCL_VAR) && (*use)->AsLclVarCommon()->Data()->OperIsConvertMaskToVector())
+        {
+            // Found
+            //      use:STORE_LCL_VAR(ConvertMaskToVector(x))
+            lclOp            = (*use)->AsLclVarCommon();
+            isLocalStore     = true;
+            removeConversion = true;
+        }
+        else if ((*use)->OperIs(GT_STORE_LCL_VAR) && !(*use)->AsLclVarCommon()->Data()->OperIsConvertMaskToVector())
+        {
+            // Found
+            //      use:STORE_LCL_VAR(x)
+            lclOp         = (*use)->AsLclVarCommon();
+            isLocalStore  = true;
+            addConversion = true;
+        }
+        else if ((*use)->OperIsConvertVectorToMask() && (*use)->AsHWIntrinsic()->Op(2)->OperIs(GT_LCL_VAR))
+        {
+            // Found
+            //      user(use:ConvertVectorToMask(LCL_VAR(x)))
+            lclOp            = (*use)->AsHWIntrinsic()->Op(2)->AsLclVarCommon();
+            isLocalUse       = true;
+            removeConversion = true;
+        }
+        else if ((*use)->OperIs(GT_LCL_VAR) && ((user == nullptr) || !user->OperIsConvertVectorToMask()))
+        {
+            // Found
+            //      user(use:LCL_VAR(x))
+            lclOp         = (*use)->AsLclVar();
+            isLocalUse    = true;
+            addConversion = true;
+        }
+        else
+        {
+            // Found something else
+            return fgWalkResult::WALK_CONTINUE;
+        }
+
+        assert(isLocalStore != isLocalUse);
+        assert(addConversion != removeConversion);
+        assert(lclOp != nullptr);
+
+        // Get the existing weighting.
+        MaskConversionsWeight weight;
+        bool                  found = weightsTable->Lookup(lclOp->GetLclNum(), &weight);
+        assert(found);
+
+        // Quit if the cost of changing is higher or is invalid.
+        if (weight.currentCost <= weight.switchCost || weight.invalid)
+        {
+            JITDUMP("Local %s V%02d at [%06u] will not be converted. ", isLocalStore ? "store" : "var",
+                    lclOp->GetLclNum(), Compiler::dspTreeID(lclOp));
+            weight.DumpTotalWeight();
+            return fgWalkResult::WALK_CONTINUE;
+        }
+
+        JITDUMP("Local %s V%02d at [%06u] will be converted. ", isLocalStore ? "store" : "var", lclOp->GetLclNum(),
+                Compiler::dspTreeID(lclOp));
+        weight.DumpTotalWeight();
+
+        // Fix up the type of the lcl and the lclvar.
+        assert(lclOp->gtType != TYP_MASK);
+        var_types lclOrigType = lclOp->gtType;
+        lclOp->gtType         = TYP_MASK;
+        LclVarDsc* varDsc     = m_compiler->lvaGetDesc(lclOp->GetLclNum());
+        varDsc->lvType        = TYP_MASK;
```

Comment:
This needs a check to skip the conversion for parameters (`lvIsParam`) and for OSR locals (`lvIsOSRLocal`). For OSR locals there may have been stores in the tier 0 version that you did not see and that you thus cannot update.

This suggests that the transformation would probably be better off by avoiding the retyping and instead creating new `TYP_MASK` locals, updating all uses to the new locals. The required handling for parameters and OSR locals would just be to insert a single initial conversion in an initial basic block. That extra conversion could be taken into account in the heuristic.

Up to you if you want to put in the restriction or change it in this suggested way. I'd probably suggest to just put in the restriction and improve it in a follow-up if you ever run into some motivating cases.

Reply:
> and instead creating new TYP_MASK locals

That would be quite a change to this PR. I would expect most uses of this pass to come from locals within a method. But, yes, it'd be fairly easy to write some tests that expose this. For now I'm happy this stay as is as it should catch almost all instances of importance. We can do some investigations once there is some real SVE code out there - there are probably more important SVE performance issues to do first.

Rephrased your comment into the code as a TODO.